### ISSUE TITLE
Sporadic crashes in grpc_core::GrpcPolledFdWindows::SendVUDP

### ISSUE THREAD

**Issue by okravets** [grpc/grpc#22555]
### What version of gRPC and what language are you using?
1.27.1  

### What operating system (Linux, Windows,...) and version?
Windows

### What runtime / compiler are you using (e.g. python version or version of gcc)
MSVC 16.4.5

### What did you do?
Application sporadically crashing 

### What did you expect to see?
Not crashing 

### What did you see instead?
Stack point to thread which calls run_some_timers()method and then  grpc_core::ExecCtx::Get()->Flush().  

 	lakecli!abort() Line 77	C++	Symbols loaded.
>	lakecli!grpc_core::GrpcPolledFdWindows::SendVUDP(grpc_core::WSAErrorContext * wsa_error_ctx, const iovec * iov, int iov_count) Line 378	C++	Symbols loaded.
 	[Inline Frame] lakecli!grpc_core::GrpcPolledFdWindows::SendV(grpc_core::WSAErrorContext *) Line 364	C++	Symbols loaded.
 	lakecli!grpc_core::SockToPolledFdMap::SendV(unsigned __int64 as, const iovec * iov, int iovec_count, void * user_data) Line 829	C++	Symbols loaded.
 	[Inline Frame] lakecli!socket_write(ares_channeldata * s, unsigned __int64 data, const void *) Line 197	C	Symbols loaded.
 	lakecli!ares__send_query(ares_channeldata * channel, query * query, timeval * now) Line 867	C	Symbols loaded.
 	[Inline Frame] lakecli!process_timeouts(ares_channeldata *) Line 561	C	Symbols loaded.
 	lakecli!processfds(ares_channeldata * channel, fd_set * read_fds, unsigned __int64 read_fd, fd_set * write_fds, unsigned __int64 write_fd) Line 132	C	Symbols loaded.
 	lakecli!ares_process_fd(ares_channeldata * channel, unsigned __int64 read_fd, unsigned __int64 write_fd) Line 153	C	Symbols loaded.
 	lakecli!on_ares_backup_poll_alarm_locked(void * arg, grpc_error * error) Line 286	C++	Symbols loaded.
 	lakecli!grpc_combiner_continue_exec_ctx() Line 237	C++	Symbols loaded.
 	lakecli!grpc_core::ExecCtx::Flush() Line 156	C++	Symbols loaded.
 	[Inline Frame] lakecli!run_some_timers() Line 134	C++	Symbols loaded.
 	lakecli!timer_main_loop() Line 237	C++	Symbols loaded.
 	lakecli!timer_thread(void * completed_thread_ptr) Line 286	C++	Symbols loaded.
 	lakecli!`anonymous namespace'::ThreadInternalsWindows::thread_body(void * v) Line 128	C++	Symbols loaded.
 	kernel32.dll!00007ffa3e327974()	Unknown	Non-user code. Symbol loading disabled by Include/Exclude setting.
 	ntdll.dll!00007ffa3edda271()	Unknown	Non-user code. Symbol loading disabled by Include/Exclude setting.
![image](https://user-images.githubusercontent.com/7624443/78276333-e57ea500-74e0-11ea-9668-051fca936dcc.png)

### Anything else we should know about your project / environment?
We saw this happening on 1.19, 1.23 and still happening

**Comment by okravets** [comment]
I could further troubleshoot it, but app is Windows service and it is tough to gather logs. 
Any advise what should settings should I set for GRPC_TRACE to get more info to troubleshoot it?

**Comment by apolcyn** [comment]
Running under `GRPC_TRACE=api,client_channel_routing,cares_resolver,cares_address_sorting` may give some insight

**Comment by lemaxw** [comment]
interested in progress, have similar issue on 1.26.0

**Comment by ggavrilov** [comment]
We have the same issue to in 1.26.0.
```
 C:\grpc\src\core\ext\filters\client_channel\resolver\dns\c_ares\grpc_ares_ev_driver_windows.cc(378) -> assertion failed: GRPC_SLICE_LENGTH(write_buf_) == 0
```

**Comment by apolcyn** [comment]
If someone has a run under environment variables: `GRPC_TRACE=api,client_channel_routing,cares_resolver,cares_address_sorting`, that could help to get some insight on this.

For even better debug info if it's possible, an accompanying packet trace (e.g. pcap file for DNS traffic) captured during the failure, would also give insight.

**Comment by okravets** [comment]
Here is a change we did to fix this issue. Haven't seen it after. 
[0001-Attempt-to-fix-crash-reported-as-sporadic-crashes-in.txt](https://github.com/grpc/grpc/files/5124660/0001-Attempt-to-fix-crash-reported-as-sporadic-crashes-in.txt)

**Comment by gmcchessney** [comment]
I'm having the same problem in version 1.32. I'm using c# with .net framework 4.7.1.
The crash seems to be related to the connection being lost to the gRPC server machine. I haven't been able to replicate it with any GRPC_TRACE information yet.

Attached is the stack trace. It's very similar to the one in the original question.

[gRPC Disconnect Crash Stack Trace.txt](https://github.com/grpc/grpc/files/5371421/gRPC.Disconnect.Crash.Stack.Trace.txt)

**Comment by gmcchessney** [comment]
@apolcyn
I managed to get some trace information for the crash. Attached are the last 100,000 lines of the log (about 1 minute before the crash).
I can provide a larger portion of the log file if that would be helpful. The full file is about 45 million lines long.

[gRPC Error trace info.zip](https://github.com/grpc/grpc/files/5446439/gRPC.Error.trace.info.zip)

**Comment by Alyona-E** [comment]
@apolcyn 
I`m able to reproduce this issue by toggling VPN state (on/off). We`re using grpc 1.28.
[grpc_assert_trace.txt](https://github.com/grpc/grpc/files/6146552/grpc_assert_trace.txt)

The issue doesn`t reproduce with the patch provided by @okravets.

**Comment by apolcyn** [comment]
Thank you all for the great info! I've merged the mentioned patch as #25726

**Comment by kartlee80** [comment]
Hi,

Which version of grpc will this be in? We are getting few reports from people and I am unable to reproduce the crash in windows - switching hibernate mode / vpn on or off as mentioned here.

I am happy to upgrade the grpc to the version once we know.

-Karthik

**Comment by okravets** [comment]
Hey ! I saw it as I mentioned in a ticket. After we've upgraded to 1.28.1-1,  I still applied my patch and do not see it.
Please see my comment from August 25, 2020.

Sasha

**Comment by apolcyn** [comment]
@okravets a fix with the patch you provided is included starting at https://github.com/grpc/grpc/releases/tag/v1.36.4 (which includes https://github.com/grpc/grpc/pull/25690)

**Comment by sali98** [comment]
> @okravets a fix with the patch you provided is included starting at https://github.com/grpc/grpc/releases/tag/v1.36.4 (which includes #25690)
> 
> So you should no longer need to apply the mentioned patch manually.

Hey @apolcyn, we are currently using grpc 1.38.1, which has contains the [25690](https://github.com/grpc/grpc/pull/25690), but we are still seeing the crash at **grpc_core::GrpcPolledFdWindows::SendVTCP**.

So far, it is one of the popular crashes for our customers, but we cannot reproduce it at office.
The callstack is as below:
`
>    grpc_core::GrpcPolledFdWindows::SendVTCP(grpc_core::WSAErrorContext * wsa_error_ctx=0x000000f3919fe6b8, const iovec * iov=0x000002d6ba9b2250, int iov_count=1) Line 392    C++
     [Inline Frame] grpc_core::GrpcPolledFdWindows::SendV(grpc_core::WSAErrorContext *) Line 347    C++
     grpc_core::SockToPolledFdMap::SendV(unsigned __int64 as, const iovec * iov=0x000002d6ba9b2250, int iovec_count=1, void * user_data) Line 786    C++
     write_tcp_data(ares_channeldata * channel=0x000002d6bad73880, fd_set * write_fds=0x0000000000000000, unsigned __int64 write_fd=13928, timeval * now=0x000000f3919fe710) Line 267    C
     processfds(ares_channeldata * channel=0x000002d6bad73880, fd_set * read_fds=0x0000000000000000, unsigned __int64 read_fd=18446744073709551615, fd_set * write_fds=0x0000000000000000, unsigned __int64 write_fd=13928) Line 131    C
     ares_process_fd(ares_channeldata * channel, unsigned __int64 read_fd, unsigned __int64 write_fd) Line 154    C
     on_writable_locked(fd_node * fdn=0x000002d6bb038080, grpc_error * error=0x0000000000000000) Line 395    C++
     [Inline Frame] std::_Func_class<void>::operator()() Line 951    C++
     grpc_core::WorkSerializer::WorkSerializerImpl::Run(std::function<void __cdecl(void)> callback={...}, const grpc_core::DebugLocation & location={...}) Line 69    C++
     grpc_core::WorkSerializer::Run(std::function<void __cdecl(void)> callback={...}, const grpc_core::DebugLocation & location={...}) Line 152    C++
     on_writable(void * arg=0x000002d6bb038080, grpc_error * error=0x0000000000000000) Line 415    C++
     exec_ctx_run(grpc_closure * closure, grpc_error * error=0x0000000000000000) Line 46    C++
     grpc_core::ExecCtx::Flush() Line 158    C++
     pollset_work(grpc_pollset * pollset=0x000002d6b652fba8, grpc_pollset_worker * * worker_hdl=0x000000f3919fead0, __int64 deadline) Line 132    C++
     cq_pluck(grpc_completion_queue * cq=0x000002d6b652fa70, void * tag=0x000000f3919feeb0, gpr_timespec deadline, void * reserved=0x0000000000000000) Line 1302    C++
     grpc_completion_queue_pluck(grpc_completion_queue * cq, void * tag, gpr_timespec deadline, void * reserved=0x0000000000000000) Line 1329    C++
     grpc::CoreCodegen::grpc_completion_queue_pluck(grpc_completion_queue * cq, void * tag, gpr_timespec deadline={...}, void * reserved=0x0000000000000000) Line 74    C++
     grpc::CompletionQueue::Pluck(grpc::internal::CompletionQueueTag * tag=0x000000f3919feeb0) Line 323    C++
     grpc::internal::BlockingUnaryCallImpl<google::protobuf::MessageLite,google::protobuf::MessageLite>::BlockingUnaryCallImpl<google::protobuf::MessageLite,google::protobuf::MessageLite>(grpc::ChannelInterface * channel=0x000002d6b3669570, const grpc::internal::RpcMethod & method, grpc::ClientContext * context=0x000000f3919ff550, const google::protobuf::MessageLite & request, google::protobuf::MessageLite * result=0x000000f3919ffa20) Line 89    C++

`

**Comment by yashykt** [comment]
@apolcyn ping

**Comment by apolcyn** [comment]
Suggest trying the latest gRPC. There were some significant changes made to this code since this issue was filed, recently: https://github.com/grpc/grpc/pull/33965

**Comment by yashykt** [comment]
I'm gonna close this assuming it was fixed in #33965 since there has been no activity since then. Please open a new issue if this still persists.
